### ISSUE TITLE
Some puzzles are not solvable

### ISSUE THREAD

**Issue by hashlash** [hashlash/prolog-battleships#7]
Check the backend! Try some problems from https://puzzlemadness.co.uk/battleships

**Comment by hashlash** [comment]
One problem is on the frontend when parsing the problem from puzzlemadness.co.uk. The representation of `shipCounts` is different.

But, after altering the `shipCounts` manually, I still found an error:

```
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 58014)
Traceback (most recent call last):
  File "/home/hashlash/.pyenv/versions/3.7.4/lib/python3.7/socketserver.py", line 316, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/home/hashlash/.pyenv/versions/3.7.4/lib/python3.7/socketserver.py", line 347, in process_request
    self.finish_request(request, client_address)
  File "/home/hashlash/.pyenv/versions/3.7.4/lib/python3.7/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "server.py", line 11, in __init__
    super().__init__(*args, **kwargs)
  File "/home/hashlash/.pyenv/versions/3.7.4/lib/python3.7/http/server.py", line 646, in __init__
    super().__init__(*args, **kwargs)
  File "/home/hashlash/.pyenv/versions/3.7.4/lib/python3.7/socketserver.py", line 720, in __init__
    self.handle()
  File "/home/hashlash/.pyenv/versions/3.7.4/lib/python3.7/http/server.py", line 426, in handle
    self.handle_one_request()
  File "/home/hashlash/.pyenv/versions/3.7.4/lib/python3.7/http/server.py", line 414, in handle_one_request
    method()
  File "server.py", line 21, in do_POST
    result = solve(ships, row_clues, col_clues, grid)
  File "/media/hashlash/data/ac/8/pemlog/battleships/battleship.py", line 23, in solve
    return next(result)['List']
  File "/media/hashlash/data/ac/8/pemlog/battleships/env/lib/python3.7/site-packages/pyswip/prolog.py", line 127, in __call__
    "Returned: '", str(term), "'."]))
pyswip.prolog.PrologError: Caused by: 'Rows=[[_, _, _, _, _, _], [_, _, _, 0, _, _], [_, _, _, 1, _, 0], [_, _, _, 1, 0, 1], [_, _, _, _, _, 0], [_, _, _, _, _, _]], setof(Rows, battleship([1, 1, 1, 2, 2, 3], [1, 2, 2, 2, 2, 1], [0, 4, 0, 4, 0, 2], Rows), List)'. Returned: 'error(resource_error(stack), dict(stack_overflow, 102, choicepoints, 9, depth, 6, environments, 158, globalused, 24, localused, [Functor(532877,3,9,:(system, $add_findall_bag([6])),[]), Functor(532877,3,8,:($bags, findall_loop([6], /(:, 2), _1616, [])),[]), Functor(532877,3,7,:(system, setup_call_catcher_cleanup(/(:, 2), /(:, 2), _1652, /(:, 2))),[]), Functor(532877,3,3,:($bags, setof([6], /(:, 2), _1694)),[]), Functor(532877,3,1,:(user, pyrun([222], [2])),[])], stack, 1048576, stack_limit, 45, trailused))'.
----------------------------------------
```

Will check on the backend.

**EDIT**: It seems the cause is stack overflow.

Below is formatted message from the last line of the error message with some annotation:

```
pyswip.prolog.PrologError: Caused by:
'Rows=[[_, _, _, _, _, _],
       [_, _, _, 0, _, _],
       [_, _, _, 1, _, 0],
       [_, _, _, 1, 0, 1],
       [_, _, _, _, _, 0],
       [_, _, _, _, _, _]],
 setof(Rows,
       battleship([1, 1, 1, 2, 2, 3],
                  [1, 2, 2, 2, 2, 1],
                  [0, 4, 0, 4, 0, 2],
                  Rows),
       List)'.
Returned: 
'error(resource_error(stack),  <--- RESOURCE ERROR: STACK
       dict(stack_overflow, 102,  <--- STACK OVERFLOW
            choicepoints, 9,
            depth, 6,
            environments, 158,
            globalused, 24,
            localused, [
                Functor(532877,3,9,:(system, $add_findall_bag([6])),[]),
                Functor(532877,3,8,:($bags, findall_loop([6], /(:, 2), _1616, [])),[]),
                Functor(532877,3,7,:(system, setup_call_catcher_cleanup(/(:, 2), /(:, 2), _1652, /(:, 2))),[]),
                Functor(532877,3,3,:($bags, setof([6], /(:, 2), _1694)),[]),
                Functor(532877,3,1,:(user, pyrun([222], [2])),[])], stack, 1048576, stack_limit, 45, trailused))'.
                                                                      ^                  ^
                                                                      |-------STACK------|
```

**Comment by hashlash** [comment]
It seems the problem is caused by Prolog's `setof/3` predicate.

I try to add a `problem/5` predicate:

```prolog
% https://puzzlemadness.co.uk/battleships/small/2020/5/19
problem(4,
        [1, 1, 1, 2, 2, 3],
        [1, 2, 2, 2, 2, 1],
        [0, 4, 0, 4, 0, 2],
        [[_, _, _, _, _, _],
         [_, _, _, 0, _, _],
         [_, _, _, 1, _, 0],
         [_, _, _, 1, 0, 1],
         [_, _, _, _, _, 0],
         [_, _, _, _, _, _]]).
```

The following run without problem

```prolog
?- problem(4, Ships, Nrow, Ncol, Rows), battleship(Ships, Nrow, Ncol, Rows).
```

But this will hang:

```prolog
?- problem(4, Ships, Nrow, Ncol, Rows), setof(Rows, battleship(Ships, Nrow, Ncol, Rows), List).
```

`setof/3` predicate was added in 'https://github.com/hashlash/prolog-battleships/pull/1/commits/594cdf8f01ca46adbf40c3a1dc936f9ab50444b2#diff-918e9fe82fb425b35dc630144d1abb77' to remove duplicated models. Without this, previously, a 4x4 grid single-solution problem can generate thousands of duplicated models. It is important to remove those duplications when we want to retrieve multiple models for problems with multiple solutions.

Any opinion about this? @mridho2828

**Comment by hashlash** [comment]
For now, I think we can just use the first model retrieved and discuss the duplication removal on a different issue.

**Comment by mridho2828** [comment]
Yes, when `setof/3` predicate is called directly using SWI-Prolog I also got the "hang" issue. However, when I started `server.py` file in localhost and then calling the POST method, it worked properly. Still no idea what cause this.

If we want to retrieve the first model, then it is sufficient to add cut `!` in the Prolog rule.

**EDIT**: The same issue occurs when calling the prolog method through `server.py` file. I think the best solution for now i we want to remove duplicate and not getting issues with prolog is to add cut `!` in the rule.
